### DISCUSSION TITLE
How to "select" TextNode?

### DISCUSSION THREAD

**Discussion by evgenyny** [ueberdosis/tiptap#3300] [General Questions] [answered]
Hey!

I have a TextNode (pic attached), which I extracted from the editor. And also I have a mark applied to this TextNode. And I want to unset this specific mark from the TextNode that I have.

Main point here is that I need to unset a mark without manual selection from the user.

I've tried a lot to create prosemirror Selection on fly and then toggle mark, but to instantiate Selection you need to provide positions, and I don't have them. And also I cannot find a way to get them somehow.

Also it's not really clear for me why TextNode is not selectable? I mean it's property https://tiptap.dev/api/schema#selectable.
Because of that I wasn't able to create NodeSelection via https://prosemirror.net/docs/ref/#state.NodeSelection^create.
Probably there are some options to make it selectable?

Finally I even tried to just create WEB API Selection from Ranges - https://developer.mozilla.org/en-US/docs/Web/API/Selection, but seems like it's not possible, seems like tiptap or prosemirror preventing me to do that.

I'm one step away from just rough update directly in the JSON of my editor's content, because I can't find other options to make it correctly through APIs.

Probably you can suggest something?

Thanks in advance!

![image](https://user-images.githubusercontent.com/22072081/195194169-00830be1-85cd-499a-af74-c3343a9d7537.png)

**Comment by anton-liubushkin** [comment, marked as answer]
I'm not sure I got it right, but maybe this will help...

To unset mark from given range:

```typescript
// 1st option with text selection
editor.chain().setTextSelection({ from: 0, to: 10 }).unsetMark( typeOrName ).run()

// 2nd option without selection
editor.chain().command(({tr}) => {
    const from = 0;
    const to = 10;
    tr.removeMark(from, to)
    // tr.removeMark(from, to, markType)
    return true
} ).run()
```

To get mark range from the given position oy can use the `getMarkRange`:

```typescript
import { getMarkRange } from "@tiptap/core";

const GIVEN_POS = 10;

const range = getMarkRange(
    editor.view.state.doc.resolve( GIVEN_POS ), 
    editor.view.state.schema.marks.link
);
```

**Reply by evgenyny** [reply]
Thank you for your answer.

Thing is I don't know position which you're using in 

`setTextSelection({ from: 0, to: 10 })`

I don't have them.

I have exact Node instance, and I know that it has exact one mark, and I want to unset this mark from that TextNode.

**Reply by anton-liubushkin** [reply]
Each "paragraph" or other node with inline content can have multiple TextNodes (a separate node for each combination of marks). 

```html
<p>My awesome text</p> // → Paragraph with 1 TextNode

<p>My <strong>awesome</strong> text</p> // → 3 TextNodes. The 2nd with a Bold mark.
   ---        -------         -----
   1          2               3
```

That is, Mark has no position in the TextNode. It applies to the whole TextNode.

You can GET all these marks, delete the ones you don't need, and create a copy of the TextNode with the new set of marks.

```typescript
// Get all marks and then filter them
const oldMarks = myOldTextNode.marks;

// Or use an empty array
const newMarks = [];

const myNewTextNode = myOldTextNode.mark( newMarks );
```

But if you want to change the content, it's better to make changes as I described above, through the position. Otherwise, you'll have to bring back the modified TextNode afterwards via replacements.

**Reply by evgenyny** [reply]
Okay, it's clear, thank you so much for your answer!
